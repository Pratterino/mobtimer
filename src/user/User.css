.user {
    display: inline-flex;
    border: 1px solid #61dafb;
}

.user__image {
    border-radius: 45%;
    width: 80px;
    height: 80px;
    background-color: #61dafb;
    background-size: cover;
    background-position: center;
}

.user__name {
    align-self: flex-end;
    text-align: center;
}

.active {
    border: 4px solid mediumspringgreen;
}
