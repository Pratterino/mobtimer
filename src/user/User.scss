@import './../Global.scss';

$imageSize: 80px;
$userSpacing: 19px;
$borderSize: 3px;

.user {
    display: flex;
    flex-direction: column;
    text-align: center;
    position: relative;
    margin: 0 $userSpacing;

    figure {
        margin: 0;
    }

    &:hover {
        .user__remove {
            visibility: visible;

            svg {
                cursor: pointer;
            }
        }

        .user__crown {
            transform: rotate(0deg);
            left: 0;
        }
    }

    .user__crown {
        background: url(../assets/images/crown.png) no-repeat;
        background-size: cover;
        height: 50px;
        width: $imageSize+10;
        top: -32px;
        left: -$userSpacing;
        transform: rotate(-20deg);
        position: absolute;
        z-index: 2;
        transition: all 0.5s;
    }

    .user__image {
        position: relative; // important for .shine to not overflow.
        border-radius: 50%;
        width: $imageSize;
        height: $imageSize;
        text-indent: 0;
        background-size: cover;
        background-position: center;
        background-repeat: no-repeat;
        border: $borderSize solid var(--stopped-timer-color);
        overflow: hidden;

        &.active {
            border: $borderSize solid var(--active-timer-color);
        }

        &.disabled {
            filter: blur(1px);
            opacity: 0.4;
        }
    }

    .user__name {
        display: block;
        position: relative;
        font-weight: normal;
        bottom: 0;
    }

    .user__remove {
        visibility: hidden;
        display: flex;
        justify-content: flex-end;

        svg {
            &:hover {
                opacity: 0.6;
                color: #f00;
            }
        }
    }

    .user__name--text {
        @extend .text-shadow;

        &.active {
            color: var(--active-timer-color);
            font-weight: bold;
        }

        &:hover {
            cursor: pointer;
        }
    }

    .user__name--text,
    .user__name--text > input {
        width: 100%;
        text-align: center;
    }

    .shine {
        animation: shine 5s ease-in-out infinite;
        animation-fill-mode: forwards;
        content: '';
        position: absolute;
        top: -110%;
        left: -210%;
        width: 200%;
        height: 200%;
        opacity: 0;
        transform: rotate(30deg);

        background: rgba(255, 255, 255, 0.13);
        background: linear-gradient(
                to right,
                rgba(255, 255, 255, 0.13) 0%,
                rgba(255, 255, 255, 0.13) 77%,
                rgba(255, 255, 255, 0.5) 92%,
                rgba(255, 255, 255, 0) 100%
        );
    }
}

/* animation */
@keyframes shine {
    10% {
        opacity: 1;
        top: -30%;
        left: -30%;
        transition-property: left, top, opacity;
        transition-duration: 0.7s, 0.7s, 0.15s;
        transition-timing-function: ease;
    }
    100% {
        opacity: 0;
        top: -30%;
        left: -30%;
        transition-property: left, top, opacity;
    }
}
