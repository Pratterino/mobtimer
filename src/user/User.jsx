import React, {Component} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import './User.css';

class User extends Component {
    render() {
        const {name, image} = this.props.user;
        const style = {backgroundImage: `url(${image})`};
        return (
            <div className="user">
                <div className="user__image" style={style}/>
                <p>{name}</p>
            </div>
        );
    }
}

const mapStateToProps = state => ({});
const mapDispatchToProps = dispatch => bindActionCreators({}, dispatch);


export default connect(mapStateToProps, mapDispatchToProps)(User);
