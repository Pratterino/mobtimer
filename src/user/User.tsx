import React, { Component, createRef } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { changeName, changeUserImage, removeUser, toggleUser } from './userActions';
import './User.scss';

interface IProps {
    user: {
        name: string,
        image: string,
        disabled: boolean,
        active: boolean,
    }
    removeUser: Function,
    toggleUser: Function,
    changeUserImage: Function,
    changeName: Function,
}

interface IInput {
    current: {
        focus: Function
    }
}

class User extends Component<IProps> {
    input = createRef<HTMLDivElement>();
    state = {
        name: this.props.user.name,
        editMode: false,
    };

    onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({
            name: event.target.value,
        });
    };

    updateName = (e:React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        this.props.changeName(this.props.user, this.state.name);
        this.enterEditMode(false);
    };

    enterEditMode = (boolean = true) => {
        this.setState({ editMode: boolean });
        if (this.input.current) {
            this.input.current.focus();
        }
    };

    render() {
        // TODO: disabled should be able to be in user, why is it not updating if it is?
        const { name, image, disabled, active } = this.props.user;
        const classes = classNames({ active, disabled });
        const backgroundImage = { backgroundImage: `url(${image})` };

        return (
            <div className="user">
                {active && <div className="user__crown" />}
                {!active && (
                    <div className="user__remove">
                        <FontAwesomeIcon
                            icon={faTimesCircle}
                            onClick={this.props.removeUser.bind(null, this.props.user)}
                        />
                    </div>
                )}
                <figure
                    className={`user__image pointer ${classes}`}
                    onDoubleClick={this.props.changeUserImage.bind(null, this.props.user)}
                    onClick={this.props.toggleUser.bind(null, this.props.user)}
                    style={backgroundImage}>
                    {active && <div className="shine" />}
                </figure>
                <figcaption className="user__name">
                    <div className={classNames('user__name--text', { active })} onClick={() => this.enterEditMode(true)}>
                        {this.state.editMode ? (
                            <input
                                ref={this.input as any}
                                value={this.state.name}
                                onChange={this.onChange}
                                onBlur={this.updateName}
                            />
                        ) : (
                            name
                        )}
                    </div>
                </figcaption>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch: Dispatch) =>
    bindActionCreators(
        {
            removeUser,
            toggleUser,
            changeName,
            changeUserImage,
        },
        dispatch,
    );

export default connect(null, mapDispatchToProps)(User);

export const unwrapped = User;
