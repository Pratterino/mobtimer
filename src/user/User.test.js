import React from 'react';
import {unwrapped as UnwrappedUser} from './User';
import {shallow} from 'enzyme';

describe("User", () => {
    let wrapper;
    let changeNameSpy;

    const renderComponent = (props) => {
        changeNameSpy = jest.fn();

        wrapper = shallow(
            <UnwrappedUser
                user={{name: "John"}}
                removeUser={jest.fn()}
                toggleUser={jest.fn()}
                changeName={changeNameSpy}
                changeUserImage={jest.fn()}
                {...props}
            />
        );
    };

    beforeEach(() => {
        renderComponent();
    });

    describe('onChange', () => {
        it('should setState', () => {
            const e = {
                target: {
                    value: "Kurt",
                }
            };
            expect(wrapper.state('name')).toEqual('John');
            wrapper.instance().onChange(e);
            expect(wrapper.state('name')).toEqual('Kurt');
        });
    });

    describe('updateName', () => {
        it('should call changeName', () => {
            const e = {
                preventDefault: jest.fn(),
            };

            wrapper.instance().updateName(e);
            expect(changeNameSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.state('editMode')).toEqual(false);
        });
    });

    describe('enterEditMode', () => {
        it('should set editMode to true by default', () => {
            wrapper.instance().enterEditMode();
            expect(wrapper.state('editMode')).toBe(true);
        });
        it('should set editMode to false', () => {
            wrapper.instance().enterEditMode(false);
            expect(wrapper.state('editMode')).toBe(false);
        });
        it('should set editMode to true', () => {
            wrapper.instance().enterEditMode(true);
            expect(wrapper.state('editMode')).toBe(true);
        });
    });

    describe('render', () => {
        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
