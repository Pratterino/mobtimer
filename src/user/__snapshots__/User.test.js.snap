// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User render should match snapshot 1`] = `
<div
  className="user"
>
  <div
    className="user__remove"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f057",
            "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z",
          ],
          "iconName": "times-circle",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </div>
  <figure
    className="user__image pointer "
    onClick={[Function]}
    onDoubleClick={[Function]}
    style={
      Object {
        "backgroundImage": "url(undefined)",
      }
    }
  />
  <figcaption
    className="user__name"
  >
    <div
      className="user__name--text"
      onClick={[Function]}
    >
      John
    </div>
  </figcaption>
</div>
`;
